### PR TITLE
SDK-as-Modules refactor + Python Support

### PR DESCRIPTION
**Pull request by sipsma**
### Goals
1. Get Python SDK-as-Module working
1. Simplify SDK-as-Module development+packaging process to extent possible
2. Setup builtin SDKs to be usable without hard dependency on internet (i.e. pulling images from dockerhub)

### Overview of modified approach
(This has lots of overlap with previous iteration, not a mutually exclusive description)

1. `dagger.json` only has an `sdk` field, which is either the name of a built-in SDK ("go", "python", etc.) OR a module reference (no longer container image references)
   * ATM just assuming they are fixed references, have TODO for re-integrating moving references
1. Go SDK is the one special SDK which is not implemented as a Module
   * The Go `codegen` binary is packaged up in a container w/ the go runtime and then included in the engine image as tarball
   * That container is loaded by the engine and used to run `Codegen` and `ModuleRuntime` implementations, coded in `core/schema/sdk.go` as the `goSDK` struct
2. All other SDKs are Modules that must implement `Codegen` and `ModuleRuntime` Functions
   * There is still a notion of "builtin" SDKs, which is intended to consist of our "official" SDKs that we bundle w/ the Engine image (to prevent hard deps on the internet, though finishing the full implementation of that requirement is a TODO)
   * Builtin SDKs besides Go are still Modules though, they just are loaded from source that is packaged in the engine image
3. An SDK-as-Module has to choose its own SDK for its Module. This can be any Module, builtin or external
    * Python chooses Go because it's the second SDK in existence and thus has no other choice
    * You can thus think of a DAG of SDK Modules, where all paths eventually lead back to Go

### TODOs:
- [x] Verify a custom module using the Python SDK actually works
   * I can successfully run `dagger mod init --sdk python --name test` as long as I have a minimal `pyproject.toml` and empty `main.py` file and obtain what looks like the expected codegen output (under `sdk/`)
   * So it seems the ModuleRuntime and Codegen functions from the PythonSDK Module are basically working
   * However, I'm not sure where to go from here as I don't know how to write a module using the Python SDK :-) cc @helderco 
   * EDIT: with Helder's help and some [fixes](https://github.com/dagger/dagger/pull/5900/commits/451f255356bcab2ee827e5eb48a5a1f964a1f090) we've got it working!
- [x] General cleanup+dedupe with older code, commenting new code more, etc.
- [x] Fix weird incongruity between builtin sdk dagger.json values and hardcoded paths in `core/schema/sdk.go`. Need to just honor `dagger.json` at all times.
- [x] Add integ-test coverage for Python
- [x] Fix include/exclude paths for non-Go sdks ([link](https://github.com/dagger/dagger/blob/63421c42e138ddf1392950a97bd476d8ad955cd9/core/schema/sdk.go#L52-L53))
- [x] Re-add use of container labels to point to paths in modules (or equivalent, have some variations on idea in mind) ([link](https://github.com/dagger/dagger/blob/63421c42e138ddf1392950a97bd476d8ad955cd9/sdk/python/runtime/main.go#L12))
   * Realized this actually already worked as is; `core` no longer has any reliance on any particular paths, it just uses the workdir and entrypoint set on the container returned by `ModuleRuntime`, so we're good to go
- [x] Include module name as inputs to Codegen/ModuleRuntime (or possibly just give them the whole pre-loaded Module config?)
   * Did this by adding `ModuleConfig` to our actual API, so SDK modules that need more metadata can load it as needed. [Commit](https://github.com/dagger/dagger/pull/5900/commits/4f3c681e20669d43c32ce72b5b284c43128a869b)
   * This feels a bit more re-usable and nicer long-term than having SDK modules implementations update the opts they accept whenever some new need arises (or worst case and that doesn't end up being a problem, it's harmless)

### Follow-ups:
- Support for pinning moving references to SDK modules
   * I am going to save this for a follow-up I think; for now we will just leave SDK module references as specified by the user. I think pinning support is good, but there should be a way to also `dagger mod sync --upgrade-sdk` or something to "unpin" and update. That starts getting way out of scope here, better for follow-up.
- Get the "no-internet" requirement fully working; need to include tarball w/ python interpreter, vendor modules/deps, etc.
   * I realized that this is probably best solved by something much more generic than just "include relevant base image for SDK". This is because built-in SDKs often need "base image + some commands + vendored deps" (i.e. `python:3.10`+`apk add git` + `pip install`)
   * The approach I have in mind is to use another optional Function on SDK Modules, `Base`, that returns a `Container` with everything needed to run the Module w/out internet deps. Then, we can add a mechanism to our CI for calling those, exporting to a local cache and including that cache in the engine image.
   * That should result in the simplest SDK dev process I think, in addition to perhaps being a useful feature outside just SDK development
   * However, this is getting way out of scope of this PR, so better for follow-up
- Fix host directory ".." restriction (or adjust working dir of modules) ([link](https://github.com/dagger/dagger/blob/63421c42e138ddf1392950a97bd476d8ad955cd9/sdk/python/runtime/main.go#L82)) ([discord discussion](https://discord.com/channels/707636530424053791/1162509681126809710/1162509683593072711))
   * This is important but also not a blocker since there's a quick workaround; lifting the `..` restriction is worth it's own discussion so I'm disentangling from this PR

### CODE REVIEW

**Comment by sipsma** [comment]
Think this is good enough to go, marked as ready for review.

I ended up trimming some stuff that can be saved for follow-ups; mentioned in the description. Also updated the description with more details on how I fixed the various TODOs when relevant, may be helpful for context.

cc @vito @helderco @jedevc

**Comment by sipsma** [comment]
I was the unlucky one to hit new trivy vuln scan failures first I think, I included fixes in this PR but then decided to spin those out to a separate one since it probably is gonna block everyone else trying to open PRs until fixed https://github.com/dagger/dagger/pull/5906

**Comment by sipsma** [comment]
@wingyplus any ideas on the elixir codegen failure here? https://github.com/dagger/dagger/actions/runs/6541674992/job/17763524119?pr=5900#step:4:1067

```
== Compilation error in file lib/dagger/gen/client.ex ==
** (KeyError) key :selection not found
    expanding struct: Dagger.ModuleConfig.__struct__/1
    lib/dagger/gen/client.ex:362: Dagger.Client.module_config/3
```

**Comment by wingyplus** [comment]
> @wingyplus any ideas on the elixir codegen failure here? https://github.com/dagger/dagger/actions/runs/6541674992/job/17763524119?pr=5900#step:4:1067
> 
> ```
> == Compilation error in file lib/dagger/gen/client.ex ==
> ** (KeyError) key :selection not found
>     expanding struct: Dagger.ModuleConfig.__struct__/1
>     lib/dagger/gen/client.ex:362: Dagger.Client.module_config/3
> ```

Let me take a look at this tonight. 🫡

**Comment by wingyplus** [comment]
@sipsma I have done some hacks on the Elixir codegen to make it pass. I have a plan to rework the Elixir codegen but it needs time to work on it. So in this way, it should not be a blocker on your PR now.

**Review by helderco — APPROVED** [review]
It's much nicer not having to depend on publishing the runtime externally, and supporting other SDKs feels pretty easy now.

LGTM! :shipit:

**Review by vito — APPROVED** [review]

File: core/modules/config.go
Comment (sipsma) [reply core/modules/config.go:1]: > Nevermind - it probably makes more sense to just add GraphQL APIs for dealing with refs so that we can limit SDK surface area and have this functionality available to all languages automatically. Especially now that the core is aware of module ref syntax via Module.sdk anyway.

Yeah that's what I was thinking; all that logic is not going to be specific to the Go SDK so may as well API-ify it. Also beneficial for any advanced users who want to write their "UX-frontends" to Dagger separate from our `dagger` CLI (alternative CLIs, web interfaces, etc.)

The nuances are well taken though, I basically just decided to trim as many decisions as possible from this PR to unblock python, but once I merge I'll pack all the follow-ups in issues so we don't lose this.

File: core/schema/container.graphqls
Comment (sipsma) [reply core/schema/container.graphqls:679]: Agreed, the multiplatform support was so long ago that I forget the exact reasoning for this exact approach, though I remember it being a practical compromise for some reason or another rather than "this is the ideal api". Would need to dig that PR up.

I have some ideas on a total re-thinking of all the publish/export/etc. related APIs in general too, so yes I agree better to just stick with this for now and save for a future wider-scale refactoring.

File: core/schema/module.graphqls
Comment (sipsma) [reply core/schema/module.graphqls:141]: Only in an integ test so far. I added this because Helder said he'll want to know the name of the module for follow-up improvements to the Python SDK module implementation. I just wanted to handle that right away to make sure it was gonna work with the new code in this PR and thus avoid another refactor or something.

For a little more context, these are the other options I could think of besides what I went with:
1. Just add `moduleName` as a parameter to the `ModuleRuntime`+`Codegen` functions
   * Felt like the beginning of an eternal game of whack-a-mole
2. Just give a single parameter `*Module` as input
   * Cool idea in that you now have a Module that accepts another Module as input, updates it and returns the modified one.
   * But less cool in that the module the SDKs accept would be in an ill-defined state of initialization, hard to understand
3. Add `ModuleConfig` to the API.
   * Could have also provided `cfg *ModuleConfig` as an input arg to the `ModuleRuntime`+`Codegen` functions, but went with this approach of putting it as an API call partially to avoid the current questions around handling of input types in the Go SDK and partially because this API seemed more potentially useful in the long run

Either way, totally agree that we should probably integrate this better into the rest of the API. `Module.config` makes sense too. Better for a follow up though

**Comment by sipsma** [comment]
Thank you for the fixes @wingyplus!

---

On a separate note, I had some commits related to the otel upgrades in here previously, but ended up splitting that out into a separate PR because it's a whole separate can of worms https://github.com/dagger/dagger/pull/5906

So I removed them from here. Provided the only CI failure remaining is the scanner, I'll override and merge. CI on main is already failing due to the scanner (we pick up new vulns "out-of-band" and thus can get failures on main at any time), so shouldn't cause any harm.

**Pull request merged**
